File: chad/src/main/java/org/woahoverflow/chad/commands/fun/UpVote.java
package org.woahoverflow.chad.commands.fun;

import java.util.ArrayList;
import java.util.List;
import org.woahoverflow.chad.framework.obj.Command;
import org.woahoverflow.chad.framework.obj.Player;
import org.woahoverflow.chad.framework.obj.Player.DataType;
import org.woahoverflow.chad.framework.handle.MessageHandler;
import org.woahoverflow.chad.framework.handle.PlayerHandler;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent;
import sx.blah.discord.handle.obj.IUser;
import sx.blah.discord.util.EmbedBuilder;

/**
 * Suppresses a player's profile
 *
 * @author sho
 * @since 0.7.0
 */
@SuppressWarnings("unchecked")
public class UpVote implements Command.Class {
    @Override
    public Runnable run(MessageReceivedEvent e, List<String> args) {
        return () -> {
            MessageHandler messageHandler = new MessageHandler(e.getChannel(), e.getAuthor());

            if (e.getMessage().getMentions().isEmpty())
            {
                messageHandler.sendError(MessageHandler.NO_MENTIONS);
                return;
            }

            // Target
            IUser target = e.getMessage().getMentions().get(0);

            // The author's player instance
            Player author = PlayerHandler.handle.getPlayer(e.getAuthor().getLongID());

            ArrayList<Long> voteData = (ArrayList<Long>) author.getObject(DataType.VOTE_DATA);

            if (voteData.contains(target.getLongID()))
            {
                messageHandler.sendError("You've already voted on this user!");
                return;
            }

            // Add the voted user to the array
            voteData.add(target.getLongID());

            // ReSet it
            author.setObject(DataType.VOTE_DATA, voteData);

            // The target user's player instance
            Player targetPlayer = PlayerHandler.handle.getPlayer(target.getLongID());

            // The target user's upvote amount
            long targetPlayerUpvote = (long) targetPlayer.getObject(DataType.PROFILE_UPVOTE);

            // The target user's downvote amount
            long targetPlayerDownvote = (long) targetPlayer.getObject(DataType.PROFILE_DOWNVOTE);

            // Update values
            targetPlayer.setObject(
                DataType.PROFILE_UPVOTE, targetPlayerUpvote+1L
                );

            messageHandler.sendEmbed(new EmbedBuilder().withDesc(
                "You upvoted `" + target.getName() + "`!\n"
                    + "Their vote is now `"+((targetPlayerUpvote+1L)-targetPlayerDownvote)+ "`!"
            ));
        };
    }

    @Override
    public Runnable help(MessageReceivedEvent e) {
        return null;
    }
}

File: chad/src/main/java/dev/shog/chad/commands/community/UpVote.kt
package dev.shog.chad.commands.community

import dev.shog.chad.framework.handle.GuildHandler
import dev.shog.chad.framework.handle.MessageHandler
import dev.shog.chad.framework.handle.PlayerHandler
import dev.shog.chad.framework.obj.Command
import dev.shog.chad.framework.obj.Guild
import dev.shog.chad.framework.obj.Player.DataType
import org.json.JSONArray
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent
import sx.blah.discord.util.EmbedBuilder
import java.util.*

/**
 * UpVote a player's profile
 *
 * @author sho
 */
class UpVote : Command.Class {
    override suspend fun run(e: MessageEvent, args: MutableList<String>) {
        val messageHandler = MessageHandler(e.channel, e.author)
        val prefix = GuildHandler.getGuild(e.guild.longID).getObject(Guild.DataType.PREFIX) as String

        if (e.message.mentions.isEmpty()) {
            messageHandler.sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, prefix + "upvote [@user]")
            return
        }

        // Target
        val target = e.message.mentions[0]

        // Makes sure they're not upvoting themselves
        if (target == e.author) {
            messageHandler.sendError("You can't vote on that user!")
            return
        }

        // The author's player instance
        val author = PlayerHandler.getPlayer(e.author.longID)

        val voteData = JSONArray(author.getObject(DataType.VOTE_DATA))

        if (voteData.contains(target.longID)) {
            messageHandler.sendError("You've already voted on this user!")
            return
        }

        // Add the voted user to the array
        voteData.put(target.longID)

        // ReSet it
        author.setObject(DataType.VOTE_DATA, voteData)

        // The target user's player instance
        val targetPlayer = PlayerHandler.getPlayer(target.longID)

        // The target user's upvote amount
        val targetPlayerUpvote = targetPlayer.getObject(DataType.PROFILE_UPVOTE) as Long

        // The target user's downvote amount
        val targetPlayerDownvote = targetPlayer.getObject(DataType.PROFILE_DOWNVOTE) as Long

        // Update values
        targetPlayer.setObject(
                DataType.PROFILE_UPVOTE, targetPlayerUpvote + 1L
        )

        messageHandler.sendEmbed(EmbedBuilder().withDesc(
                "You upvoted `" + target.name + "`!\n"
                        + "Their vote is now `" + (targetPlayerUpvote + 1L - targetPlayerDownvote) + "`!"
        ))
    }

    override suspend fun help(e: MessageEvent) {
        val st = HashMap<String, String>()
        st["upvote [@user]"] = "Upvotes a user's Chad profile."
        Command.helpCommand(st, "UpVote", e)
    }
}
